include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td

```diff
@@ -49,6 +49,12 @@ def FModuleLike : OpInterface<"FModuleLike", [Symbol, PortList, InstanceGraphMod
     InterfaceMethod<"Get the module's enabled layers.",
       "ArrayRef<Attribute>", "getLayers">,
 
+    InterfaceMethod<"Set the module's enabled layers.",
+      "void", "setLayersAttr", (ins "::mlir::ArrayAttr":$layers)>,
+
+    InterfaceMethod<"Set the module's enabled layers",
+      "void", "setLayers", (ins "::llvm::ArrayRef<Attribute>":$layers)>,
+
```

Comment:
I've been reluctant to add this generally as only `FModuleOp`/`FExtModuleOp` can have this. It seems wrong to then define this for everything and then have ops that implement the interface either not actually implement it (really bad), assert on the member function (also bad), or not put it in the interface and work with the ops manually (cumbersome).

@darthscsi or @dtzSiFive: Do you have any bright ideas here?

Reply:
Nothing bright, this is IMO one fundamental challenge for designing these things and interfaces.

Sometimes generic methods aren't reasonable for concrete ops and yet not only is it cumbersome but can be easily missed if we constantly enumerate the ops that currently interact with a certain feature and this makes for harder to read and especially hard to maintain code.

To me this is one of the more important concerns -- how to make it easy to add/change things without auditing the entire code base each time.  


Of course we can splinter functionality into more interfaces.  This probably makes sense for some bigger interfaces that combine too many things.  I think I would prefer more granular interfaces sometimes, it's not easy to draw the line re:trade-offs here.

I think/wonder if code working with Layers is distinct enough (in the best of ways) that actually just having that be its own interface for layer-ness makes sense?

Another angle is we combine ModuleOp, ExtModuleOp, and probably IntModuleOp -- as the first two are arguably the same thing just declaration/definition anyway and the last is basically an extmodule with the intrinsic marker on it or thereabouts -- as they have more in common than different (maybe) -- "isDeclaration()" already present because it's a symbol, and "intrinsic-ness" is either a bit or presence of "which intrinsic" (cc defname as it shouldn't need both).  The idea is combine the things that are the same to reduce the things we put on the interface to capture this reuse in a way that doesn't make sense for other things "like" that ("this makes sense for the things MORE like it but not the ones we want under this umbrella too").

------

I think Interfaces end up often being read-only (or return mutable range / accessor thing) to help address this sort of thing -- could something like that work here? Probably if you're generically operating at the interface level for ops that means you necessarily have one constructed and you want to mutate it's information about this aspect (here, array of layers or so) to the extent such exists.

And specifically here could probably dodge this a little -- if we wanted -- with a "clearLayers()".

Ideas but nothing great, thanks for asking hopefully we can push on this more soon :+1: .